server/jetstream_api.go

```diff
@@ -1033,6 +1084,195 @@ func (s *Server) jsStreamPurgeRequest(sub *subscription, c *client, subject, rep
 	s.sendAPIResponse(c, subject, reply, string(msg), s.jsonResponse(resp))
 }
 
+// Request to restore a stream.
+func (s *Server) jsStreamRestoreRequest(sub *subscription, c *client, subject, reply string, msg []byte) {
+	if c.acc == nil {
+		return
+	}
+	acc := c.acc
+
+	var resp = JSApiStreamRestoreResponse{ApiResponse: ApiResponse{Type: JSApiStreamRestoreResponseType}}
+	if !acc.JetStreamEnabled() {
+		resp.Error = jsNotEnabledErr
+		s.sendAPIResponse(c, subject, reply, string(msg), s.jsonResponse(&resp))
+		return
+	}
+	if !isEmptyRequest(msg) {
+		resp.Error = jsNotEmptyRequestErr
+		s.sendAPIResponse(c, subject, reply, string(msg), s.jsonResponse(&resp))
+		return
+	}
+	stream := streamNameFromSubject(subject)
+	if _, err := acc.LookupStream(stream); err == nil {
+		resp.Error = &ApiError{Code: 400, Description: fmt.Sprintf("stream [%q] already exists", stream)}
+		s.sendAPIResponse(c, subject, reply, string(msg), s.jsonResponse(&resp))
+		return
+	}
+
+	tfile, err := ioutil.TempFile("", "jetstream-restore-")
+	if err != nil {
+		resp.Error = &ApiError{Code: 500, Description: "jetstream unable to open temp storage for restore"}
+		s.sendAPIResponse(c, subject, reply, string(msg), s.jsonResponse(&resp))
+		return
+	}
+
+	// Create our internal subscription to accept the snapshot.
+	restoreSubj := fmt.Sprintf(jsRestoreDeliverT, stream, nuid.Next())
+
+	// FIXME(dlc) - Can't recover well here if something goes wrong. Could use channels and at least time
+	// things out. Note that this is tied to the requesting client, so if it is a tool this goes away when
+	// the client does. Only thing leaking here is the sub on strange failure.
+	acc.subscribeInternal(c, restoreSubj, func(sub *subscription, c *client, _, _ string, msg []byte) {
+		if len(msg) == 0 {
+			tfile.Seek(0, 0)
+			// TODO(dlc) - no way right now to communicate back.
+			acc.RestoreStream(stream, tfile)
+			tfile.Close()
+			os.Remove(tfile.Name())
+			c.processUnsub(sub.sid)
+			return
+		}
+		// Append chunk to temp file.
+		tfile.Write(msg)
```

Comment:
need to catch errors here @derekcollison

Reply:
also the lack of ack here is a bit of a problem, there's no way for a restore to know its happening/working until after its potentially sent gigs of data and then seeing the stream doesnt exist, so we need to do something better here i think